[RFC 0036] Improving the RFC process (#36)
NixOS/rfcs: rfcs/0036-rfc-process-team-amendment.md

RFC from community members nominated in the discussion on that RFC.

This team should be people who are very familiar with the main components
touched by the RFC.

##### Shepherd Leader
The Shepherd Leader is in charge of the RFC process for a specific RFC, and is
responsible for ensuring the process is followed in a timely fashion. 

## Process from Creation to Merge

*In short, to get a major change included in Nix or nixpkgs, one must
first get the RFC merged into the RFC repository as a markdown file under the
`accepted` directory. At that point the RFC is accepted and may be implemented
with the goal of eventual inclusion into Nix or nixpkgs.*

0. Have a cool idea!
1. Find a co-author. A co-author is critical to making sure your RFC is viable
   and will receive support. Your co-author helps flesh out the RFC, and should
   also support the RFC.
2. Fill in the RFC. Put care into the details: RFCs that do not present
   convincing motivation, demonstrate understanding of the impact of the design,
   or are disingenuous about the drawbacks or alternatives tend to be
   poorly-received. You might want to create a PR in your fork of the RFCs

>    report to help you flesh it out with a few supporters or chat/video

Reviewer: ```suggestion
   repo to help you flesh it out with a few supporters or chat/video
```

---
feature: rfc-process-team-amendment
start-date: 2018-10-27
author: Robin Gloster <mail@glob.in>
co-authors: Graham Christensen <graham@grahamc.com>
related-issues: 1 (initial process), 24 (implementation)
---

# Summary
[summary]: #summary

This RFC proposes an RFC Steering Committee who decide on a group of RFC
shepherds for each RFC who guide the discussion to a general consensus and then
propose a motion for a "Final Comment Period" (FCP) with a disposition for
acception, rejection or postponing. (see Glossary for a short definition of

> these terms)

Reviewer 2: I think that «postponing» is never defined, and «FCP» and «disposition» are not mentioned in the Glossary/Terminology section.

Reviewer 2: I think the idea of postponing is still unclear.

Maybe it would be better just to consider postponing a rejection, and say explicitly that new context or new tools or new knowledge can justify submitting a later RFC similar to an earlier rejected one?

Reviewer 3: Even though the idea is more or less the same, I think separating postponement and rejection is useful to avoid strong wording, that can have negative effects on both sides of the discussion (people not wanting to use it and unconsciously preferring to let a RFC stall, people feeling rejected). IMO postponement is “we want that but are not ready for it” (eg. a RFC to make nixpkgs use `builtins.encryptString` everywhere), and rejection is “as of now, we don't think we want that” (eg. a RFC to remove overlays).

Now, agree with you that they are currently not well-defined-enough in the `Terminology` section and that it could use some wording to make it explicit that RFC choices can be revisited if the context|tools|knowledge changes.

Reviewer 2: I am in favour of postponement (and of encouraging the shepherds to specify their personal — non-unanimous — expectations of what has to happen first for a real decision); but if there is a reason not to define it (or its consequences), I think omitting it might be still better than having an unclear hanging notion.

# Motivation
[motivation]: #motivation

A lot of RFCs have stalled and already an [RFC has been submitted exactly on
this topic](https://github.com/NixOS/rfcs/pull/18), which ironically has not
been decided on either. This new RFC takes the above into account and tries to

> expand on that to flesh out the process further. During this effort a lot of

Reviewer 2: Does it make any prescription for preexisting proposals (except closing #18 as superseded by the current one, I guess)?

Author: If this is accepted we will go through the past RFCs that have been decided on and follow the process as defined in the RFC. (i.e. #18 will most probably be either closed before we reach that point or else probably should be a rather easy one to reach consensus on closing)

inspiration has been taken from [Rust's RFC
process](https://github.com/rust-lang/rfcs#what-the-process-is) which works well
and we have adapted to our needs.

# Detailed design
[design]: #detailed-design

> ## Glossary

Reviewer 4: A Glossary is usually a section that is at the end of the document and that contains definitions that can be "glossed" over. For example the FCP acronym is introduced in the middle of the document, of one skiped it, it's useful to be able to go to the end and quickly look it up.

Author: We moved it to the top as we had complaints that it comes too late in the text, but the naming of that probably can be improved. Any ideas?

##### RFC Steering Committee
A team of people defined by _this_ RFC and stays consistent until the team
members are changed via a follow-up RFC. This committee is responsible for
forming an RFC Shepherd team from the available nominations on each RFC. This
team also names the leader of the Shepherd team. This has to happen within 1
week after the PR has been opened. Until then the Steering Committee is
responsible for guiding the discussion. In case of the Shepherding Team not

> doing its work the Steering Committee shall encourage them or step in and assign

Reviewer 5: "doing its work" is as of this RFC not super well defined. IMO this can evolve over time and be specified as things settle down, but it needs to be understood that *someone* (presumably the steering committee) is ultimately responsible for the meta-task of keeping things running by some process or other.

Author: I think I've addressed this.

new Shepherds. They also are in charge of merging accepted and rejected RFCs.
Generally by these expectations they should find time to meet once a week for

> about an hour.

Reviewer 6: I think we should add a paragraph about how to handle replacing members of the steering committee, in case they're not able to participate in the meetings. This would, ideally, include guidelines on _when_ to replace members: for example, if they stop showing up.

> about an hour.

Reviewer 4: It's important to clearly outline the responsibilities of the team. In it's own paragraph or bulleted list.

If we want this to work we need the members to be accountable for something.

Author: I think I've addressed this.

> ##### Shepherd

Reviewer 7: ```suggestion
##### Shepherd Team
```

> ##### Shepherd Team

Reviewer 4: wording: "Team" is associated to a durable thing whereas "Commitee" is more temporary.

* RFC Steering Committee -> RFCs Team
* RFC Sheperd Team -> RFC Sheperd Commitee

Reviewer 2: I think many conferences have Steering Committees which are responsible for long-term continuity (and they are indeed durable).

A team of 3-4 community members defined unanimously by the RFC Steering

> Committee, responsible for accepting or rejecting RFCs. This team is created per

Reviewer 7: ```suggestion
Committee, responsible for accepting or rejecting a specific RFC. This team is created per
```

RFC from community members nominated in the discussion on that RFC.

This team should be people who are very familiar with the main components

> touched by the RFC.

Reviewer 4: same, what are the responsibilities of the RFC sheperds?

> ##### Leader

Reviewer 7: ```suggestion
##### Shepherd Leader
```

The Shepherd Leader is in charge of the RFC process for a specific RFC, and is
responsible for ensuring the process is followed in a timely fashion. 

## Process from Creation to Merge

*In short, to get a major change included in Nix or nixpkgs, one must
first get the RFC merged into the RFC repository as a markdown file under the
`accepted` directory. At that point the RFC is accepted and may be implemented
with the goal of eventual inclusion into Nix or nixpkgs.*

0. Have a cool idea!

> 1. Find a co-author. A co-author is critical to making sure your RFC is viable

Reviewer 5: Author and co-author should maybe be recused from steering team and ineligible from shepherding team... Which is a shame because I'd really like to nominate @globin and @grahamc for this shepherding team :grin:

Author: Yes, I thought about that when opening the PR, but haven't had the time of adding it, yet.

   and will receive support. Your co-author helps flesh out the RFC, and should

>    also support the RFC.

Author: One can still write up the initial draft on one's own, but then before PR'ing a co-author should be found, who will probably have some remarks to flesh it out further.

>    also support the RFC.

Reviewer 2: That's true; but watching RFC repository was previously a reasonable way to notice the discussion on still-raw RFCs. Now it will spill in all the different directions…

>    also support the RFC.

Reviewer 5: I don't have a strong opinion here, but FWIW I've never quite understood why a co-author is critical nor, in the case where there is clearly a single driving force, what the co-author is actually expected to do.

>    also support the RFC.

Reviewer 8: I also don’t quite understand the motivation behind the co-author. Is it just to signal that more than a single person believe the suggestion is a good idea? Or is it more about increasing the quality of the initial draft by having a second pair of eyes looking over it before a bigger audience gets to review it? 

Depending on the actual motivation I wonder if it might end up being just a formality without too much added value?

>    also support the RFC.

Reviewer 4: The co-author was introduced in the initial RFC process to play some of the role of the sheperd's team; help shape the RFC and clean it before submitting the RFC for wider consumption. There was also the idea that if the author can't find at least one co-author then it will be hard motivating more people that the RFC is worth considering.

>    also support the RFC.

Reviewer 4: Please vote with thumbs on this comment: remove the co-author from the RFC

>    also support the RFC.

Author: @grahamc and I decided to keep it mainly for the reason of keeping up the standard of opened PRs. For poorly written our thought out RFCs that probably won't be accepted, it would add unnecessary work for the RFC SC. A second set of eyes would hopefully ensure that there is support of at least a few people and obvious fallacies can be addressed before a larger audience is let loose on it.

>    also support the RFC.

Reviewer 9: An idea: replace requirement for the co-author with "find at least one person that agrees to be nominated as part of the shepherding Team"?

So you have confirmation that at least one other person considers this important. Also, a possible shepherding member is usually more experienced so can give advice during rfc draft process.

>    also support the RFC.

Author: Removed the necessity for a co-author and fleshed out the request to the author, to gather feedback at an early stage/before opening the PR.

2. Fill in the RFC. Put care into the details: RFCs that do not present
   convincing motivation, demonstrate understanding of the impact of the design,
   or are disingenuous about the drawbacks or alternatives tend to be
   poorly-received. You might want to create a PR in your fork of the RFCs
   report to help you flesh it out with a few supporters or chat/video

>    conference with a few people involved in the topic of the RFC.

Reviewer 5: :100: In general, we as a community *seriously* underutilize synchronous voice/video calls.

3. Submit a pull request. As a pull request the RFC will receive design feedback
   from the larger community, and the author should be prepared to revise it in
   response.
4. For the nomination process for potential members of the RFC Shepherd Team,
   that is specific to each RFC, anyone interested can either nominate another

>    person or themselves to be a potential member of the RFC Shepherd Team. This

Reviewer 5: Presumably other-nominees have to accept?

Reviewer 10: I think it's clear that one shouldn't get into a shepherd team without agreeing with that :-)  We could write it down explicitly, but it could be thought as implicitly being part of step 5 (i.e. SC choosing reasonably).

   drawbacks are shown. You can make edits, big and small, to the RFC to clarify
   or change the design, but make changes as new commits to the pull request,
   and leave a comment on the pull request explaining your changes.
   Specifically, do not squash or rebase commits after they are visible on the
   pull request.
9. At some point, a member of the RFC Shepherd Team will propose a "motion for
final comment period" (FCP), along with a disposition for the RFC (merge, close,
or postpone).
    * This step is taken when enough of the tradeoffs have been discussed that
      the RFC Shepherd Team is in a position to make a decision. That does not
      require consensus amongst all participants in the RFC thread (which is
      usually impossible). However, the argument supporting the disposition on
      the RFC needs to have already been clearly articulated, and there should
      not be a strong consensus against that position outside of the RFC
      Shepherd Team. RFC Shepherd Team members use their best judgment in taking
      this step, and the FCP itself ensures there is ample time and notification

>       for stakeholders to push back if it is made prematurely.

Reviewer 5: Might be good to reference https://tools.ietf.org/html/rfc7282 as a guideline, or even a formal expectation.

Reviewer 10: Yes, it won't hurt to at least link what we mean by "consensus", and I believe this particular link is a good one.  I expect longer-term participants from a similar community to mostly have a very similar notion of what it means, but still...

Reviewer 5: Fine leaving this for now, but would be worth returning to if/when we hit more contentious RFCs

    * For RFCs with lengthy discussion, the motion to FCP is usually preceded by
      a summary comment trying to lay out the current state of the discussion
      and major tradeoffs/points of disagreement.
    * Before actually entering FCP, all members of the RFC Shepherd Team must

>       sign off the motion.

Reviewer 2: Maybe add something like:

«
If the Shepherd Committee reaches unanimous agreement that the discussion is not producing new major substantial points, but cannot reach unanimous agreement about the disposition, it is normally advised that an FCP motion of postponing the RFC due to unresolved disagreement is proposed. It can be adopted by unanimous agreement of either the Shepherd committee or the Steering Committee.
»

10. The FCP lasts ten calendar days, so that it is open for at least 5 business
days. It is also advertised widely, e.g. in NixOS Weekly and through Discourse
announcements. This way all stakeholders have a chance to lodge any final
objections before a decision is reached.
11. In most cases, the FCP period is quiet, and the RFC is either merged or
closed. However, sometimes substantial new arguments or ideas are raised, the
FCP is canceled, and the RFC goes back into development mode.
12. In case of acceptance, the RFC Steering Committee merges the PR into the

> `accepted`, in case of rejection into the `rejected` directory.

Reviewer 2: What happens in case of postponing?

Maybe in case of postponing all the active participants of the discussion (including the authors and the Shepherd Committee) are encouraged to share if there is any example of evidence that would surprise them and decrease their confidence in the current position?

![RFC Process](./0036-rfc-process.png)
![Review Process](./0036-review-process.png)

## The RFC life-cycle

Once an RFC is accepted the authors may implement it and submit the feature as a
pull request to the Nix or nixpkgs repo. Being accepted is not a rubber stamp,
and in particular still does not mean the feature will ultimately be merged; it
does mean that in principle all the major stakeholders have agreed to the

> feature and are amenable to merging it.

Reviewer 5: Should we explicitly state that a positive RFC switches the default to "no reply means no go" to "no reply means go ahead?" i.e. if some feature goes through the RFC process, then the implementation is PR'd and sits for a reasonable amount of time with the appropriate owners pinged but no substantive negative feedback and otherwise stalled conversation, should anyone with the appropriate commit bit feel free to merge?

Reviewer 2: I think if we expect such a change of perception, it needs to be explicit in the RFC. If this amendment gets direct objections, at least we will know what is the situation…

Reviewer 10: I'd expect the implementer to @mention the corresponding shepherd team, and as a result we should have at least one review of the implementation.  RFCs are supposed to be about more significant changes, and I wouldn't like to merge those without at least a second pair of eyes on the code changes.

Yes, there might be some exceptions, but as today &ndash; it ultimately depends on the judgement of the one who pushes/merges a change without any reaction.  For those "trivial" RFC implementations, it should be easy enough to get another opinion like _"I haven't really looked, but this kind of change seems OK without review"._

Reviewer 5: > it ultimately depends on the judgement of the one who pushes/merges a change without any reaction.

I think our community has shown time and time again that that judgment will almost always err in favor of inaction. We're a respectful and cautious bunch, if we don't explicitly say somewhere that it's OK to go ahead in most cases we won't. If that's the desired outcome that's one thing, but if the expectation is that an accepted RFC will generally mean that a sound PR will get merged, then we need to be clear about that.

Reviewer 10: > [...] the expectation is that an accepted RFC will generally mean that a sound PR will get merged [...]

I agree with the expectation, though it's hard to say how strong it should be.  I'd say it might be individual for each RFC, judged from the discussion around it.  Sometimes it's difficult to anticipate all important information _before_ making an implementation and yet you may want an RFC to gather some more formal opinion before undertaking full implementation.

The current text does express something like what you write, though I expect you'd prefer a stronger wording:

> Being accepted is not a rubber stamp, and in particular still does not mean the feature will ultimately be merged; it does mean that in principle all the major stakeholders have agreed to the feature and are amenable to merging it.

Reviewer 2: I would say that given the project's track record on reaching a decision or at least finding out the reason why decision has not been reached, this wording doesn't really promise that RFCs mean anything.

I would have more hopes for a text like 

> Of course, an implementation PR for an accepted RFC can still be rejected due to implementation-specific concerns, or due to some substantial issues not discussed in the context of RFC that become apparent in the context of an implementation.

I guess that would be too optimistic:

> If substantial new (or implementation-specific) issues are not raised in a reasonable time, it would be acceptable to merge for any willing person with commit access after a single positive detailed implementation review (by a person who didn't participate in the implementation).

Reviewer 9: Maybe there should be a deadline for review in the implementation PR, just like there is for the RFC PR? The sheperds could be responsible for this: say a implementation PR musst enger final review after x days of no new comments, and then must either be closed or merged after 10 days. This should prevent stalled PRs at least.

Author: We haven't added that on purpose, since the discussion should continue without time limit as long as it is constructive, new points are brought up and the RFC is iterated on. If this is the case no longer, then the Shepherd Team shall step in with a motion for FCP.

completion: authors should not expect that other project developers will take on
responsibility for implementing their accepted feature.

Minor modifications to accepted RFCs can be done in follow-up pull requests. We
strive to write each RFC in a manner that it will reflect the final design of
the feature; but the nature of the process means that we cannot expect every
merged RFC to actually reflect what the end result will be after implementation.

In general, once accepted, RFCs should not be substantially changed. Only very
minor changes should be submitted as amendments. More substantial changes should
be new RFCs, with a note added to the original RFC. Exactly what counts as a
"very minor change" is up to the RFC Shepherd Team of the RFC to be amended, to
be decided in cooperation with the RFC Steering Committee.

## Members of the RFC Steering Committee

In cooperation and discussion with Eelco Dolstra and all nominees the proposal
for the first iteration of members of the RFC Steering Committee are:

 - Eelco Dolstra (edolstra, niksnut)
 - Shea Levy (shlevy)
 - Domen Kožar (domenkozar)
 - Jörg Thalheim (Mic92)
 - Robin Gloster (globin)

> 

Reviewer 4: For convenience, it would be good to start mapping teams to github teams. Eg: @NixOS/rfc-steering . Not a blocker but I would like to see this happening.

Related, I should be removed write access on the rfcs repo since I haven't be selected in the initial team.

Reviewer 5: Agreed about github teams (though naturally that should wait until after this RFC is accepted, if indeed it is).

Ambivalent about the commit bit question. So long as we trust everyone with the appropriate permissions to follow the relevant policies (including not accidentally running afoul of them), I don't think we need to be overly strict on access control.

Author: I agree. Also as @shlevy has said I don't necessarily thing that your commit bit should be removed. I don't see you trying to cause any harm etc.

# Drawbacks
[drawbacks]: #drawbacks

If the Steering Committee were too biased, it might select a biased Shepherding
Team. We are hoping for them and believe them to commit to doing their work in
the interest of the community. Also this RFC introduces more process and
bureaucracy, and requires more meetings for some core Nix/nixpkgs contributors.
Precious time and energy will need to be devoted to discussions.

# Alternatives
[alternatives]: #alternatives

> The current state, which hardly ever results in an RFC being accepted.

Reviewer 5: Other alternatives include well defined owners for any particular domain, with the rights to decide on the process relevant to their domain and the responsibility to follow through on that. An extreme example oft his case is a BDFL with responsibility for all final decisions.

Reviewer 2: Well defined owners plus a process for recognition of new domains, probably. That process can be either well defined or indeed explicitly be a judgement call by a leader.

Author: I'll try to add that

Reviewer 10: I don't think that's really a "hard alternative".  I expect that SC will be only glad for anything that makes the shepherd choice straightforward.  We have CODEOWNERS now already, although perhaps under-used, and the file probably can't really cover completely everything, even if we have it in every repo.

To be clear, I do believe that we need to move a bit in direction of maintainership/ownership (at least in nixpkgs).

# Unresolved questions
[unresolved]: #unresolved-questions

None, as of now.

# Future work
[future]: #future-work

Work on auto-labeling RFCs and automation of parts of the process that either do
not need human intervention or to remind people to continue their work.

> Define how the Steering Committee is picked in the future.

Reviewer 5: A lot of formalization of steering and shepherding responsibilities should be fleshed out as we go IMO.

new Shepherds. They also are in charge of merging accepted and rejected RFCs.
Generally by these expectations they should find time to meet once a week for
about an hour.

They have no special responsibility with regard to the content of an RFC, they
can weigh in on them, the same as any other community member, but are only in
charge of:
 * selecting the Shepherds unanimously
 * supervising that the Shepherds are carrying out their work
 * committing the final RFC

##### Shepherd Team
A team of 3-4 community members defined unanimously by the RFC Steering
Committee, responsible for accepting or rejecting a specific RFC. This team is
created per RFC from community members nominated in the discussion on that RFC.

This team should be people who are very familiar with the main components
touched by the RFC. The author cannot be part of the Shepherd Team. In addition,
at most half of the Shepherd Team can be part of the RFC Steering Committee.

The resposibility of the team is to guide the discussion as long as it is
constructive, new points are brought up and the RFC is iterated on and from time
to time summarise the current state of discussion. If this is the case no longer,

> then the Shepherd Team shall step in with a motion for FCP.

Reviewer 2: Is the main decision point  accept/reject the FCP motion disposition? Maybe it would be a good idea then to add that the Shepherd Team also has to unanimously decide, after studying all the points raised during the discussion, whether to recommend accepting the RFC?
